Add flux-pgrep and flux-pkill

grondo opened the pull request:
This PR adds a minimal version of `flux-pgrep`/`flux-pkill`. This allows matching and canceling jobs by name and/or jobid range without the more advanced match/constraint syntax proposed in #4849. Once RFC 35 in flux-framework/rfc#360 is approved, then we can extend pgrep to support the more advanced syntax once again.

This version of `flux-pgrep` supports matching a pattern  to job names by default. Due to user request, a jobid range is also supported if the match expression appears as `ID1..ID2`. Since that is also a valid pattern, a name regex can be forced via `name:<pattern>`.

grondo commented on src/cmd/flux-pgrep.py:
```diff
@@ -0,0 +1,281 @@
+
+@flux.util.CLIMain(LOGGER)
+def main():
+    sys.stdout = open(sys.stdout.fileno(), "w", encoding="utf8")
```
This is a false positive, we don't need to close stdout.

garlick commented on doc/man1/flux-pgrep.rst:
```diff
@@ -0,0 +1,98 @@
+By default, only active jobs for the current user are considered.
+
+*flux-pkill* cancels matching jobs instead of listing them. This is
+equivalent to running *flux-grep* with the ``-k`` option.
```
s/grep/pgrep/

garlick commented on src/cmd/flux-pgrep.py:
```diff
@@ -0,0 +1,284 @@
+##############################################################
+# Copyright 2019 Lawrence Livermore National Security, LLC
```
update copyright date

garlick reviewed (commented):
couple superficial comments from a quick first pass.

Good stuff!

garlick commented on src/cmd/Makefile.am:
```diff
@@ -85,7 +85,9 @@ dist_fluxcmd_SCRIPTS = \
 	flux-uri.py \
-	flux-pstree.py
+	flux-pstree.py \
+	flux-pgrep.py \
+	flux-pkill.py
```
When installed, the symlink is lost and two copies of the executable are installed.   In other projects I've seen `install-exec-hook` used to create symbolic links, as described in the autotools docs:

https://www.gnu.org/software/automake/manual/automake.html#Extending

I'm not sure if it's worth the extra complexity honestly, but thought I'd mention it.

grondo replied:
Thanks I looked into that when doing the original `pgrep` PR but I can't remember why I discarded it. I'll take another look. FWIW, the main `pgrep`/`pkill` package also installs a copy and not a symlink...

grondo replied:
Took another look and I think I've got this working. It is unfortunately quite a bit more complex since you need both an install and uninstall-local hook, plus a rule to generate the flux-pkill.py symlink locally so that `flux pkill` works in-tree. However, we don't then have to commit a symlink to git so I guess that's better?

garlick replied:
Sorry about that, I just wasn't sure what the current best practice was for that sort of thing.  Whichever way makes you happy is OK with me.

grondo commented:
That can't work because then `flux job cancel` would have to query the list of jobs.

chu11 commented:
> That can't work because then flux job cancel would have to query the list of jobs.

Oh duh!  yeah, this has to be specific to `pgrep/pkill` and only it, users can't expect it to "go to" the other tools.

chu11 commented on doc/man1/flux-pgrep.rst:
```diff
@@ -0,0 +1,98 @@
+
+Options for ``flux-pgrep`` only:
+
+**-o, --format**\ *=NAME|FORMAT*
+   Specify a named output format *NAME* or a format string using Python's
+   format syntax. For more information see the documentation in
+   :man1:flux-jobs. Supported named formats include *default*, *full*,
+   *long*, and *deps*.
```
thought: given you probably wanted to leave `-a` close to `-A` above, instead of having a `pgrep only` section, maybe just note "for pgrep only" for those options that are pgrep only?

grondo replied:
As noted below: Good idea!

chu11 commented on t/t2811-flux-pgrep.t:
```diff
@@ -0,0 +1,65 @@
+	flux pgrep -a $(head -n 1 ids)..$(tail -n1 ids2) >pgrep4.ids &&
+	test $(wc -l <pgrep4.ids) -eq 12
+'
+test_expect_success 'flux-pgrep assumes regex with invalid jobid(s)' '
+	flux pgrep sl..p
```
should add a test for the special char `:`?

grondo commented on t/t2811-flux-pgrep.t:
```diff
@@ -0,0 +1,65 @@
+export SHELL=/bin/sh
+
+test_expect_success 'flux-pgrep errors on invalid arguments' '
+	test_must_fail flux pgrep &&
+	test_must_fail flux pgrep foo bar &&
+	test_must_fail flux pgrep x1..x2 &&
+	test_must_fail flux pgrep f1..f10..f13
```
Yeah, you caught me being lazy ;-)

grondo commented on doc/man1/flux-pgrep.rst:
```diff
@@ -0,0 +1,98 @@
+OPTIONS
+=======
+
+**-a**
+   Include jobs in all states, including inactive jobs.
+   This is shorthand for *--filter=pending,running,inactive*.
```
No this is for `pkill` also, but was there something that led you to believe `-a` was for `pgrep` only?

chu11 replied:
```
   parser.add_argument(
        "-a",
        action="store_true",
        help="Target jobs in all states"
        if PROGRAM == "flux-pgrep"
        else argparse.SUPPRESS,
    )
```

you suppress the option if its pgrep (unless I'm reading this wrong, I assume black formatted things on different lines)

grondo replied:
Oh, yeah, duh, that makes sense (I was thinking -A) since you can only cancel active jobs :facepalm: I like your idea of marking options `pgrep-only`

grondo replied:
I wonder if `--filter` should also be suppressed in `flux-pkill`? (Sorry, I hadn't looked at this portion of the code since before vacation, so thanks for catching that one!)

grondo replied:
I guess filter still may be useful to cancel only pending jobs for example.

chu11 replied:
i think keeping `--filter` is a good idea for both.

chu11 replied:
> Oh, yeah, duh, that makes sense (I was thinking -A) since you can only cancel active jobs facepalm I like your idea of marking options pgrep-only

Noticed the way pgrep(1) does it is:

```
       -signal
       --signal signal
              Defines the signal to send to each matched process.  Either the numeric or the symbolic signal name can be  used.   (pkill only.)

       -d, --delimiter delimiter
              Sets the string used to delimit each process ID in the output (by default a newline).  (pgrep only.)
```

which I sort of like.

grondo replied:
Yeah, agreed. I'll update the man page to use that style.

chu11 replied:
it just occurred to me, SUPPRESS hides the option, doesn't remove it.  was that intentional?  I guess if someone did `-a` with `flux-pkill` it's not the end of the world, although we'd be trying to cancel a bunch of inactive jobs?

grondo replied:
This option needs to use SUPPRESS because `args.a` is used in the code when fetching jobs, otherwise Python would throw an exception there that `a` doesn't exist in the `Namespace`. We could add a check that `args.a` isn't set when `PROGRAM == "flux-pkill"` but that didn't seem worth it for now, and it occurs to me that might be useful in testing.

chu11 commented on doc/man1/flux-pgrep.rst:
```diff
@@ -0,0 +1,98 @@
+**--max-entries**\ *=N*
+   Limit the number of jobs to consider to *N* entries (default 1000).
+
+**-k, --kill**
+   Cancel matching jobs instead of printing them. Equivalent to *flux-pkill*.
```
yeah, maybe dropping it would just make things simpler, since `pgrep` doesn't have a `--kill` option.

chu11 commented:
just realized, there's not `--signal` option for `flux pkill`, necessary?   Could create an issue and maybe for another day if users request?

grondo commented:
> just realized, there's not --signal option for flux pkill, necessary? Could create an issue and maybe for another day if users request?

Yeah, I thought about that, but was trying to reduce the complexity at this point (and the amount of required testing). A user that needs to signal  a set of jobs could always do `flux job kill -s SIGNAL $(flux pgrep PATTERN)`

grondo commented:
I just squashed and force-pushed an update to address all or most of @chu11's great comments (and fix an issue with the generation of the `flux-pkill.py` symlink)

 - dropped the `flux pgrep --kill` option as suggested
 - reworked docs to use _pgrep_ only/_pkill_ only in options as necessary, instead of a section for `pgrep` options
 - updated tests as suggested

chu11 commented on doc/man1/flux-pgrep.rst:
```diff
@@ -0,0 +1,95 @@
+
+*flux-pgrep* lists jobids that match a supplied expression. The
+expression may contain a pattern which matches the job name, or
+a range of jobids in the form ``jobid1..jobid2``. If both a pattern
+and jobid range are supplied then both must match.
```
based on this sentence, should there be a test for pattern & jobid range?

chu11 reviewed (approved):
overall LGTM, just the minor comment below about one extra potential test and see other comment I made about if SUPPRESS was the right/intended choice for `-a`.

grondo commented on src/cmd/flux-pgrep.py:
```diff
@@ -0,0 +1,281 @@
+
+@flux.util.CLIMain(LOGGER)
+def main():
+    sys.stdout = open(sys.stdout.fileno(), "w", encoding="utf8")
```
Why do I have to keep dismissing this alert Github?!?

grondo commented:
I'll set MWP if someone approves that `flux-pgrep` now shares configuration with `flux-jobs`.

chu11 commented:
LGTM!
